src/crypto/qat/qcccrypto.cc

```diff
@@ -322,150 +359,221 @@ bool QccCrypto::perform_op(unsigned char* out, const unsigned char* in,
+  sessionSetupData.cipherSetupData.pCipherKey = pCipherKey;
+  sessionSetupData.cipherSetupData.cipherDirection = cipherDirection;
+
+  if (nullptr == *sessionCtx) {
+    status = cpaCySymDpSessionCtxGetSize(cyInstHandle, &sessionSetupData, &sessionCtxSize);
+    if (likely(CPA_STATUS_SUCCESS == status)) {
+      status = qcc_contig_mem_alloc((void **)(sessionCtx), sessionCtxSize);
+    } else {
+      dout(1) << "cpaCySymDpSessionCtxGetSize failed with status = " << status << dendl;
     }
   }
+  if (likely(CPA_STATUS_SUCCESS == status)) {
+    status = cpaCySymDpInitSession(cyInstHandle,
+                                   &sessionSetupData,
+                                   *sessionCtx);
+    if (unlikely(status != CPA_STATUS_SUCCESS)) {
+      dout(1) << "cpaCySymDpInitSession failed with status = " << status << dendl;
+    }
+  } else {
+    dout(1) << "Session alloc failed with status = " << status << dendl;
+  }
+  return status;
+}
 
-  // Copy src & iv into the respective buffers
-  memcpy(qcc_op_mem[avail_inst].src_buff, in, size);
-  memcpy(qcc_op_mem[avail_inst].iv_buff, iv, AES_256_IV_LEN);
+template <typename CompletionToken>
+auto QatCrypto::async_perform_op(int avail_inst, std::vector<CpaCySymDpOpData*>& pOpDataVec, CompletionToken&& token) {
+  CpaStatus status = CPA_STATUS_SUCCESS;
+  using boost::asio::async_completion;
+  using Signature = void(std::error_code);
+  async_completion<CompletionToken, Signature> init(token);
+  auto ex = boost::asio::get_associated_executor(init.completion_handler);
+  completion_handler = [this, ex, handler = init.completion_handler](std::error_code ec) {
+    boost::asio::post(ex, std::bind(handler, ec));
+  };
+
+  status = cpaCySymDpEnqueueOpBatch(pOpDataVec.size(), &pOpDataVec[0], CPA_TRUE);
+
+  if (status != CPA_STATUS_SUCCESS) {
+    auto ec = std::error_code{status, std::generic_category()};
+    completion_handler(ec);
+  }
+  return init.result.get();
+}
 
-  //Assign the reminder of the stuff
-  qcc_op_mem[avail_inst].src_buff_flat->dataLenInBytes = qcc_op_mem[avail_inst].buff_size;
-  qcc_op_mem[avail_inst].src_buff_flat->pData = qcc_op_mem[avail_inst].src_buff;
+bool QccCrypto::symPerformOp(int avail_inst,
+                              CpaCySymSessionCtx sessionCtx,
+                              const Cpa8U *pSrc,
+                              Cpa8U *pDst,
+                              Cpa32U size,
+                              Cpa8U *pIv,
+                              Cpa32U ivLen,
+                              optional_yield y) {
+  CpaStatus status = CPA_STATUS_SUCCESS;
+  Cpa32U one_batch_size = chunk_size * MAX_NUM_SYM_REQ_BATCH;
+  Cpa32U iv_index = 0;
+  for (Cpa32U off = 0; off < size; off += one_batch_size) {
+    QatCrypto helper(this);
+    std::vector<CpaCySymDpOpData*> pOpDataVec;
+    for (Cpa32U offset = off, i = 0; offset < size && i < MAX_NUM_SYM_REQ_BATCH; offset += chunk_size, i++) {
```

Comment:
> it looks like this vector has a known maximum size of MAX_NUM_SYM_REQ_BATCH. we might avoid dynamic allocations here by using `boost::container::static_vector<CpaCySymDpOpData*, MAX_NUM_SYM_REQ_BATCH>` instead
> 
> this detail could be hidden from `async_perform_op()` by passing it as a `std::span<CpaCySymDpOpData*>` instead of its vector type

Done.